### ISSUE TITLE
H2 segfaults when switching to next song while receiving MIDI

### ISSUE THREAD

**Issue by thijz** [hydrogen-music/hydrogen#661]
Hydrogen version *          : be8ecd073b167e569d7a61b881d3a9505014f5eb
Operating system + version  : ubuntu 16.04
Audio driver + version      : jack

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
when you use a playlist you can hit F5/F6 to jump to the next /previous song in the playlist
if you hit F5/F6 while you are sending MIDI to H2 (via a controller interface of soft midi controller) H2 segfaults
this is 100% reproducible

**Comment by mauser** [comment]
Hm.. it segfaults here also without any midi activity while switching songs :smile:

**Comment by mauser** [comment]
I've fixed one rather obvious bug when removing a song, but i'm not sure if this was also your problem which is described in this issue. Could you re-test this problem with the current master?

**Comment by thijz** [comment]
just retested with the latest code : problem remains
will gather some more data and post it here

**Comment by thijz** [comment]
running h2 in gdb gives me this when h2 segfaults : 


> Thread 12 "hydrogen" received signal SIGSEGV, Segmentation fault.
> [Switching to Thread 0x7fffc4e1e700 (LWP 12518)]
> 0x0000000000505f4a in H2Core::Song::get_instrument_list (this=0x0) at /home/tvanseve/Code/prive/hydrogenMay2018/src/core/include/hydrogen/basics/song.h:130
> 130				return __instrument_list;

**Comment by elpescado** [comment]
Changing song involves unloading old drumkit and replacing it with one freshly loaded from new song. Maybe Hydrogen receives MIDI message while no drumkit is loaded?

Best regards,
Przemek

**Comment by thijz** [comment]
Ok i'm in trouble :-(
My BCR controller is acting up and 2 knobs just jitter more and more : they randomly jump up/down 1 without me touching the knobs and thus send midi to H2 at random times. 
This is what made me log this issue, but it has gotten wors, a lot worse) ... as a consequence it is now nearly impossible to switch between songs using F5/F6.
To make matters worse i have a gig this weekend and i have no time to get a new BCR + config it properly (i have a highly customized config on the BCR) ... 
Would there be an easy fix for this issue ?

I am trying to see if i can temporarily disconnect the BCR from H2 during a switch to the next song, but it's not that simple to do so all help is highly appreciated !

**Comment by mauser** [comment]
Hi! 

There is a chance that this issue does not happen with 1.0-beta1, so you could try to use that version. Maybe i can have a look at the issue tomorrow, but i can't promise a quick fix.

**Comment by elpescado** [comment]
Hi,

I looked into this issue - I hope my fix solves this issue.

However, I discovered potential larger issue. It seems that MIDI drivers are [active objects](https://en.wikipedia.org/wiki/Active_object) i.e. they process incoming MIDI message in their own thread, either directly, by calling `pthread create` (ALSA driver) or indirectly, by underlying libraries (CoreMIDI and JACK). That means that MIDI messages are handled in separate thread which means that Hydrogen state is being manipulated by two threads (main thread and MIDI handler thread) - that might lead to synchronization issues.

I imagine that solution would be that MIDI drivers push incoming MIDI messages into `EventQueue` and main thread would pop them from there and handle them. I'm unsure if that's possible, how much effort would be required and what would be performance impact though.
